iam-token-manager/v1.ts

```diff
@@ -0,0 +1,198 @@
+   */
+  public requestToken(authHeader: string, cb: Function): void {
+    const authorization = authHeader || 'Basic Yng6Yng=';
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: authorization
+        },
+        form: {
+          grant_type: 'urn:ibm:params:oauth:grant-type:apikey',
+          apikey: this.iam_apikey,
+          response_type: 'cloud_iam'
+        }
+      }
+    };
+    sendRequest(parameters, (error, body, response) => {
+      cb(error, body);
+    });
+  }
+
+  /**
+   * Refresh an IAM token using a refresh token.
+   *
+   * @param {string} authHeader - The authorization header to send to the service. Defaults to 'Basic Yng6Yng='.
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public refreshToken(authHeader: string, cb: Function) {
+    const authorization = authHeader || 'Basic Yng6Yng=';
+    const parameters = {
+      options: {
+        url: this.url,
+        method: 'POST',
+        headers: {
+          'Content-type': 'application/x-www-form-urlencoded',
+          Authorization: authorization
+        },
+        form: {
+          grant_type: 'refresh_token',
+          refresh_token: this.tokenInfo.refresh_token
+        }
+      }
+    };
+    sendRequest(parameters, (error, body, response) => {
+      cb(error, body);
+    });
+  }
+
+  /**
+   * Set a self-managed IAM access token.
+   * The access token should be valid and not yet expired.
+   *
+   * By using this method, you accept responsibility for managing the
+   * access token yourself. You must set a new access token before this
+   * one expires. Failing to do so will result in authentication errors
+   * after this token expires.
+   *
+   * @param {string} iam_access_token - A valid, non-expired IAM access token
+   * @returns {void}
+   */
+  public setAccessToken(iam_access_token: string): void { // tslint:disable-line variable-name
+    this.userAccessToken = iam_access_token;
+  }
+
+  /**
+   * Check if currently stored token is expired.
+   * 
+   * Using a buffer to prevent the edge case of the 
+   * token expiring before the request could be made.
+   *
+   * The buffer will be a fraction of the total TTL. Using 80%.
+   *
+   * @private
+   * @returns {boolean}
+   */
+  private isTokenExpired(): boolean {
+    const fractionOfTtl = 0.8;
+    const timeToLive = this.tokenInfo.expires_in;
+    const expireTime = this.tokenInfo.expiration;
+    return expireTime - (timeToLive * (1.0 - fractionOfTtl))
+      < Math.floor(Date.now() / 1000);
+  }
+
+  /**
+   * Save the response from the IAM service request to the object's state.
+   *
+   * @param {IamTokenData} tokenResponse - Response object from IAM service request
+   * @private
+   * @returns {void}
+   */
+  private saveTokenInfo(tokenResponse: IamTokenData): void {
```

Comment:
This is a synchronous request so we can use positional parameters

Reply:
The IAM service returns a single object that I am passing into this function to be set to the object state. It seems unnecessary to me to destruct the object into variables, pass them in as positional parameters, and re-construct the object to save in the state. What do you think?

Reply:
I was just giving an example of when to you positional parameters. You method was fine 😋